### PR TITLE
Send SNI in all TLS connections opened by Squid

### PR DESCRIPTION
**Pull request by rustybird**
... if possible.

Add SNI support to regular (i.e., not SslBumped) TLS connections,
including transparent http->https rewrite and Secure ICAP.

### CODE REVIEW

**Comment by rousskov** [comment]
OK to test.

The Semaphore CI check [failure](https://semaphoreci.com/squid/squid/branches/pull-request-123/builds/1) will probably go away after the PR branch is updated to be in sync with master.

**Comment by rustybird** [comment]
> The Semaphore CI check [failure](https://semaphoreci.com/squid/squid/branches/pull-request-123/builds/1) will probably go away after the PR branch is updated to be in sync with master.

Oops, sorry. I've force pushed the rebased commit just now.

**Comment by rousskov** [comment]
@chtsanti, please review this PR and my previous comment.

**Comment by rustybird** [comment]
> I think this PR can be accepted without significant changes because it fixes a problem, and while the fix is of poor quality, producing a high-quality fix is going to be difficult.

I have to agree with your characterization of my patch.

> However, if you can, please disregard my vote and refactor all SSL_set_ex_data(ssl_ex_index_server)-calling code by adding a new PeerConnector::setServer(serverAddress) method [...]

Apparently, if `url_rewrite_host_header` has been turned off, PeekingPeerConnector will currently use the original hostname (from before the rewrite) in the SNI, but use the rewritten hostname to verify the certificate? Not sure if that's really the case and if it should be preserved - I'm totally unfamiliar with SSL bumping. (Which makes me kind of nervous about refactoring security-related code.)

There's also the question, should this effect of `url_rewrite_host_header` on SNI be replicated in BlindPeerConnector? And where to get the original hostname from at that point - fish it out of the HTTP request headers? But what if the client didn't include a `Host` header in its request?

**Comment by rousskov** [comment]
You are asking many good questions. Unfortunately, I cannot answer all those questions without spending the time I do not currently have on researching those issues. Fortunately, I hope that the answers are not relevant here: My review does not suggest polishing how the server name is calculated/obtained, only what happens after that. I certainly agree that there are probably problems with how the server name is calculated in various context as well, but it would be unfair to ask you to fix those problems IMO. Your changes may make those problems more visible/apparent, which is already a step in the right direction.

In other words, I am only asking to consolidate the code that gets engaged after the name is obtained (from various semi-inconsistent sources). Some code refactoring may be necessary to separate name extraction from "Is it an IP address?" test, but nothing major, I hope.

**Comment by rustybird** [comment]
> In other words, I am only asking to consolidate the code that gets engaged after the name is obtained

The problem (if I understand the side effects of `url_rewrite_host_header` correctly) is that in PeekingPeerConnector, there isn't just _one_ server name: The server name for certificate validation - `SSL_set_ex_data(serverSession.get(), ssl_ex_index_server, (void*)hostName)` - can differ from the server name for SNI.

But I've pushed a new new commit series, which at least factors out the raw IP check. It does indeed make the mismatch between the two server names more apparent. A one-liner also added SNI support  to IcapPeerConnector.

File: src/ssl/support.cc [hunk 1]
```diff
@@ -926,6 +926,11 @@ Ssl::verifySslCertificate(Security::ContextPointer &ctx, CertificateProperties c
 bool
 Ssl::setClientSNI(SSL *ssl, const char *fqdn)
 {
+    URL tmpUrl;
```
Comment (rousskov) [inline src/ssl/support.cc]: A URL object is rather expensive to create. Ideally, the "is this an IP address?" test should be factored out, but doing so correctly is difficult, so let's just use what hostIsNumeric() (and other code) does. Something like this may work:

```
const Ip::Address test(fqdn);
if (!test.isAnyAddr())
    return; // raw IP is inappropriate for SNI
```

File: src/ssl/support.cc [hunk 2]
```diff
@@ -926,6 +926,11 @@ Ssl::verifySslCertificate(Security::ContextPointer &ctx, CertificateProperties c
...
+    tmpUrl.host(fqdn);
+    if (tmpUrl.hostIsNumeric())  // raw IP not appropriate for SNI
+        return true;
```
Comment (rousskov) [inline src/ssl/support.cc]: Please adjust Ssl::setClientSNI() return type from `bool` to `void`:

* Nobody is checking the return value right now.
* It is much simpler to not specify the return type than to make sure that a `true` return value for an essentially invalid `fqdn` parameter is what all future callers would want/expect.
* If errors become important to future callers, then the function should probably throw on errors instead (TODO).

Do not forget to update the method description as well.

**Comment by yadij** [comment]
url_rewrite_host_header should only affect the HTTP Host header sent to the server (long after the handshake and verification completes). Its only purpose is to allow badly configured or written backend servers to continue operating without bug fixes - by causing conflicts between Host header and URL being fetched.

The SNI for HTTP(S) at least should always be the URL domain/host name. IMO the cert should also be that server, but may not be. Hence maybe the "url_rewrite_host_header off" use in the first place.

That said, it may be a better idea not to send SNI at all if url_rewrite_host_header OFF is configured. That way we minimize the amount of conflicting information sent to a server already known to be confused about what domain it is providing service for.

**Comment by rustybird** [comment]
@rousskov:
> I only have a few minor polishing requests [...]

All done in the newest commit series, thanks!

@yadij:
> The SNI for HTTP(S) at least should always be the URL domain/host name.

Not sure about that: The [commit](https://github.com/squid-cache/squid/commit/c52a46930129466e6c0ae3a5c82eadeb59bb463f) for the logic that sets SNI to the original hostname on `url_rewrite_host_header off` fixed [bug 4293](https://bugs.squid-cache.org/show_bug.cgi?id=4293). Which talks about rewriting `www.foo.com` to `www.foo-internal.com`, so it makes sense to me. But the part I don't get is, why not go even further and validate the certificate as `www.foo.com` then, in that use case. Anyway, I'm probably the wrong person to touch that logic.

**Comment by rustybird** [comment]
Force pushed again - I hadn't properly changed the stubbed out version of `setClientSNI()` to `void` as well.

**Comment by rousskov** [comment]
> I'm probably the wrong person to touch that logic.

To be more precise, this is probably the wrong PR to touch that logic! Besides host rewriting questions (driven by the `redirected` condition), there are also questions related to the `!hostName` condition -- it is not obvious to me why we may want to leave hostName (i.e., ssl_ex_index_server) nil but still send SNI. The hostName-setting code talks about cases where "we may not know the server name that the client will request" but (a) those use cases seem to apply to SNI-sending code as well and (b) there are other cases where hostName remains nil, so the difference between hostName and SNI is not clear to me.

> Not sure about that: The [commit](https://github.com/squid-cache/squid/commit/c52a46930129466e6c0ae3a5c82eadeb59bb463f) for the logic that sets SNI to the original hostname on `url_rewrite_host_header off` fixed [bug 4293](https://bugs.squid-cache.org/show_bug.cgi?id=4293). Which talks about rewriting `www.foo.com` to `www.foo-internal.com`, so it makes sense to me. But the part I don't get is, why not go even further and validate the certificate as `www.foo.com` then, in that use case.

Thank you for researching and articulating this! Did you mean to write "and validate the certificate as `www.foo-internal.com` then"? If Squid sends a foo-internal SNI, it should expect a foo-internal certificate, right?

**Comment by chtsanti** [comment]
I believe that the SNI should match the name we are going to use to verify server certificate which means SSL_set_ex_data(..., ssl_ex_index_server,... ) and setClientSNI should use the same name.
The code inside Ssl::PeekingPeerConnector::initialize was and it is still problematic.

About redirecting code inside this method, why not supporting redirection when peeking or staring?
In the other hand is it valid to allow redirection on bump-server-first or bump-client-first bumping modes?
Maybe in all of these cases it is better to just use only the SNI from client request and only when this info exist.

About applyTlsDetailsToSSL, it sets the SNI, only if the client uses SNI (which means details->serverName is not NULL) and uses only the client SNI, because it tries to mimic client features and behaviour. it is used only in peek or stare modes.
My sense is that this function should moved from bio.cc file and be a PeekingPeerConnector method.

**Comment by rustybird** [comment]
>> The [commit](https://github.com/squid-cache/squid/commit/c52a46930129466e6c0ae3a5c82eadeb59bb463f) for the logic that sets SNI to the original hostname on `url_rewrite_host_header off` fixed [bug 4293](https://bugs.squid-cache.orgshow_bug.cgi?id=4293). Which talks about rewriting `www.foo.com` to `www.foo-internal.com`, so it makes sense to me. But the part I don't get is, why not go even further and validate the certificate as `www.foo.com` then, in that use case.

> Did you mean to write "and validate the certificate as `www.foo-internal.com` then"? If Squid sends a foo-internal SNI, it should expect a foo-internal certificate, right?

No, I really meant it should probably use `www.foo.com` for validation (that's what Squid sends as SNI).

In the setup from the ticket (which I interpret as a webserver that's only aware of its public-facing name `www.foo.com`, but is actually reached by Squid through `www.foo-internal.com`), it would seem to make sense to use `www.foo.com` for all three things - SNI, validation, and `Host` header. But currently Squid will use `www.foo-internal.com` for validation.

I'd guess that the bug reporter manually configured `tls_outgoing_options domain=www.foo.com` or `tls_outgoing_options DONT_VERIFY_DOMAIN`. That could explain why it worked for them.

**Comment by chtsanti** [comment]
@rustybird 
> There's also the question, should this effect of url_rewrite_host_header on SNI be replicated in BlindPeerConnector?

Nope this is does not make sense for BlindPeerConnector. 
Actually the related code inside PeekingPeerConnector forces squid, when the client-first bumping mode is used (or bump on step1), to not use the client SNI (client-to-squid side SNI) as SNI when connecting to the tls web server, even if this info exist but instead use as SNI the hostname from request URL  (which maybe it is rewritten). 

@rousskov @yadij 
I am not sure that the patch applied for  bug 4293 is OK. I am not sure that the bug is a bug at all. Because squid in bumping mode  is normal to send the SNI which is sent by the client.
Also the code fixed the bug4293 affects only client-first bumping mode (or bump at step1). 
If it is correct, should we implement it for other bumping modes too? But in this case we need to check for the effects of a such change.

**Comment by rousskov** [comment]
It sounds like the SNI/hostName mismatch and bug 4293 is a large can of worms. If this PR does not break things "as is", then I suggest committing it "as is" (because it clearly improves things) while also filing a bug report and try to fix the related code (because it will require more discussions/work).

@rustybird, please feel free to declare all the discussion about SNI/hostName mismatch and bug 4293 out of this PR scope. If you want to endure it, we can try to fix all those problems here, but it would be unfair for us to _insist_ that you do it, and it may be wrong to do it in this PR.

**Comment by rustybird** [comment]
>> There's also the question, should this effect of url_rewrite_host_header on SNI be replicated in BlindPeerConnector?

> Nope this is does not make sense for BlindPeerConnector.

A case could be made for allowing someone to request `http://public` and have Squid transparently rewrite it to `httpS://internal` - where the final webserver is only aware of its `public` name, and `internal` might be some circuitous route to the server, such as a localhost listener (which e.g. can't be the subject of a valid certificate) forwarding the stream through an untrusted network via a custom protocol. But I'm not going to make that case, because implementing it would mean more work for me. ;)

**Comment by rustybird** [comment]
> @rustybird, please feel free to declare all the discussion about SNI/hostName mismatch and bug 4293 out of this PR scope.

[I declare!](https://www.youtube.com/watch?v=C-m3RtoguAQ)

**Comment by rousskov** [comment]
I plan to merge this if the tests pass and @chtsanti or @yadij  do not cast a negative vote.

**Comment by rousskov** [comment]
AFAICT, there is rough consensus here that if Squid sends SNI, then it should use the same host name to validate the received certificate. If somebody wants to implement that "simple" logic proactively, they are welcome to post a dedicated PR (with an explanation of how it would affect bug 4293). Otherwise, we will wait for the first bug caused by the current "sometimes the two names are different" complex logic to pop up; whoever fixes that bug should consider implementing the "simple" logic as the solution.

There are also other aspects of the problem such as not sending SNI in some cases and computing SNI specially in some other cases. There is not enough information to declare a consensus on those, but they can wait for a specific real-world use case or a bug report.
